contracts/bridge/MiniChefV21.sol

```diff
@@ -0,0 +1,444 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.6.12;
+pragma experimental ABIEncoderV2;
+
+import "@boringcrypto/boring-solidity/contracts/libraries/BoringMath.sol";
+import "@boringcrypto/boring-solidity/contracts/BoringBatchable.sol";
+import "@boringcrypto/boring-solidity/contracts/BoringOwnable.sol";
+import "./libraries/SignedSafeMath.sol";
+import "./interfaces/IRewarder.sol";
+
+/// @notice This is a modification of MiniChefV2, a SYN distribution
+/// contract, that doesn't need to have minting rights for SYN token.
+/// List of modifications:
+/// 1. It is no longer possible to add the same LP token twice.
+/// 2. Pools are updated before the new one is added.
+/// 3. IRewarder is now tracking "old user LP token balance in MiniChef"
+///    instead of "new user LP token balance in MiniChef"
+///         [see deposit() for reasoning]
+contract MiniChefV21 is BoringOwnable, BoringBatchable {
```

Comment:
Shouldn't this implement IMiniChefV2?

Reply:
@trajan0x post-compile task which takes an XOR of the function signatures of the Interface and Contract?

Reply:
> Structs defined in MinichefV2 have to be used from IMinichefV2 and cannot be redeclared. Imo this makes it a little less explicit.

These could be defined in a separate library file and imported by both the interface and the contract

Reply:
Could be interesting: https://docs.openzeppelin.com/contracts/3.x/utilities#introspection

Reply:
Nevermind. ERC1820 doesn't do what I thought it does.

After implementing, it turns out it's a registry where contracts represent they implement an interface rather then an interface checker itself. This kind of introspection would need to be done at the abi level

Reply:
Btw, looks like this contract does not in fact implement IMinichefV2. Place this in `testing`:

```
// SPDX-License-Identifier: MIT

pragma solidity 0.6.12;
pragma experimental ABIEncoderV2;

import {IMiniChefV2} from "../interfaces/IMiniChefV2.sol";
import {MiniChefV21} from "../MiniChefV21.sol";
import "@boringcrypto/boring-solidity/contracts/interfaces/IERC20.sol";

// @dev this is a compile time assetion that minichef v21 implements i minichef
contract MiniChefV21ImplementsIMinichef {
    IMiniChefV2 minichef;

    constructor(IERC20 testToken) public {
        minichef = new MiniChefV21(testToken);
    }
}
```

Reply:
Takes about 10 seconds to implement, too.

<img width="612" alt="Screen Shot 2022-03-01 at 16 22 25" src="https://user-images.githubusercontent.com/88872576/156251107-09d1fbfc-14ec-49d2-83a4-de41bda483c6.png">
<img width="519" alt="Screen Shot 2022-03-01 at 16 22 35" src="https://user-images.githubusercontent.com/88872576/156251109-774bb158-0ab4-470f-8890-7cd9f26e3ed0.png">

Reply:
@bleaker0x yeah but that doesn't actually prove anything. You're certifying that you implement the interface. ERC165 doesn't actually do any introspection

Reply:
<img width="580" alt="Screen Shot 2022-03-01 at 16 26 52" src="https://user-images.githubusercontent.com/88872576/156251723-fcfbaee4-2785-4f06-908f-b17015120447.png">

Can also do this

Reply:
Ultimately there's no good way to implement an interface without actually implementing it. 

Unfortunately the UserInfo struct defined in IMiniChefV2 uses a uint256 for `rewardDebt`, where MiniChefV21 uses an int256. IMiniChefV2 isn't actually implemented by the main MiniChefV2 contract, so this could actually just be changed

Reply:
Yeah - easiest way to go seems to be to actually implement it, even if it's less explicit

Reply:
@trajan0x https://github.com/synapsecns/synapse-contracts/pull/90

Reply:
It seems that no existing contracts are using `IMiniChefV2` interface. Therefore, I think it can be changed to be fully compatible with existing `MiniChefV2`: using uint256 for `rewardDebt` (it can't be negative anyway). This will not break any existing code, while making the new code a bit cleaner.
@trajan0x @bleaker0x

Reply:
> It seems that no existing contracts are using `IMiniChefV2` interface. Therefore, I think it can be changed to be fully compatible with existing `MiniChefV2`: using uint256 for `rewardDebt` (it can't be negative anyway). This will not break any existing code, while making the new code a bit cleaner. @trajan0x @bleaker0x

Ignore that, I'm being dumb here. `rewardDebt` in fact needs to be int256 (can be both positive and negative).